model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientModelCriteriaBuilder.java

```diff
@@ -0,0 +1,131 @@
+import java.util.stream.Stream;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientModel.SearchableFields;
+import org.keycloak.models.map.storage.jpa.hibernate.jsonb.JsonbType;
+import org.keycloak.models.map.storage.CriterionNotSupportedException;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.storage.SearchableModelField;
+
+public class JpaClientModelCriteriaBuilder implements ModelCriteriaBuilder<ClientModel, JpaClientModelCriteriaBuilder> {
+
+    private BiFunction<CriteriaBuilder, Root<JpaClientEntity>, Predicate> predicateFunc = null;
+
+    public JpaClientModelCriteriaBuilder() {
+    }
+
+    private JpaClientModelCriteriaBuilder(BiFunction<CriteriaBuilder, Root<JpaClientEntity>, Predicate> predicateFunc) {
+        this.predicateFunc = predicateFunc;
+    }
+
+    private void validateValue(Object[] value, int expectLength, SearchableModelField field, Operator op) {
+        if (value == null || value.length != expectLength) {
+            throw new CriterionNotSupportedException(field, op, "Invalid argument: " + Arrays.toString(value));
+        }
+    }
+
+    @Override
+    public JpaClientModelCriteriaBuilder compare(SearchableModelField<? super ClientModel> modelField, Operator op, Object... value) {
+        switch (op) {
+            case EQ:
+                if (modelField.equals(SearchableFields.REALM_ID) || 
+                    modelField.equals(SearchableFields.CLIENT_ID) ||
+                    modelField.equals(SearchableFields.ENABLED)) {
+
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(root.get(modelField.getName()), value[0])
+                    );
+                } else if (modelField.equals(SearchableFields.SCOPE_MAPPING_ROLE)) {
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.isTrue(cb.function("@>",
+                            Boolean.TYPE,
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fScopeMappings")),
+                            cb.literal("\"" + value[0].toString().replace("\"", "\\\"") + "\"")))
+                    );
+                } else if (modelField.equals(SearchableFields.ALWAYS_DISPLAY_IN_CONSOLE)) {
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fAlwaysDisplayInConsole")), 
+                            cb.literal(value[0]))
```

Comment:
The `value[0]` can be misused for SQL injection.
Please introduce a method to validate the type and / or escape every value used in `literal` method.
Perhaps the `validateValue` could take one more parameter - the required types?